sql: optimize the prepared statements cache

mgartner opened the pull request:
#### querycache,prep: move PrepareMetadata to new prep package

This is a mechanical change that moves `querycache.PrepareMetadata` to a
new `prep` package in preparation for future changes.

Release note: None

#### sql,prep: move PreparedStatement to prep package

This is a mechanical change the moves `sql.PreparedStatement` to the
`prep` package in preparation for future changes.

Release note: None

#### prep: add Cache

A new `prep.Cache` data structure has been added which maps prepared
statement names to `*prep.Statement`s. `Cache` provides a concise API
and it is optimized for the most common interactions in most workloads:

  1. Fetching prepared statements with `Get`.
  2. Snapshotting the `Cache` with `Commit` when no changes have been
     applied since the last `Commit`.

In order to facilitate automatic transaction retries which rewind the
command buffer, changes to the prepared statements cache must be undone.
We currently achieve this using two maps: one up-to-date map and the
other a snapshot of the prepared statements after the last transaction
committed. Before retrying, we clear the up-to-date copy of the map and
copy the contents of the snapshot into it, effectively undoing and
changes to the cache in the current transaction. If the transaction is
successful, we clear the snapshot and copy the up-to-date map into it,
effectively creating a new snapshot.

This approach is expensive even if the prepared statements cache is not
altered. If the cache hasn't changed, we can avoid clearing and copying
into the snapshot, but we still need to check for map equality to
determine if there have been changes. This equality check has time
complexity proportional to the number of cached prepared statements.

Also, there are two other maps we use implement an LRU eviction policy.
They keep track of the ordering of most recent access for each statement
(one is up-to-date and the other is a snapshot). Every time a prepared
statement is executed that was not the previously executed statement,
the most recent access ordering changes. Therefore, we can rarely avoid
the clear and copy of the snapshot like we can with the other maps.

The current implementation also lacks a clear API - the four maps are
separate objects and directly accessed across the `sql` pacakge and
prone to misuse.

`Cache` achieves similar behavior more efficiently with a clear API. It
provides `Add`, `Get`, `Remove`, and other methods with obvious purpose.
The `Rewind` and `Commit` methods facilitate automatic retries. `Commit`
is a constant-time operation if there have been no changes to the
`Cache` since the last `Commit`. See the inline comments for more
details.

The behavior of `Cache`, when used, will vary slightly from the current
behavior. Prior to this commit prepared statements were evicted as soon
new statements were added. Now, prepared statements are evicted when the
cache is committed, i.e., when statements succeed (or fail without a
retriable error) and we know that the command buffer will not be
rewound. This difference in behavior should have no meaningful impact
for users.

Fixes #143732

Release note (performance improvement): Prepared statements are now more
efficiently cached.

#### sql: replace prepared statement maps with prep.Cache

Four maps using to implement the prepared statements cache have been
repalced with `prep.Cache`. Additional, a snapshot of a connection's
portals is now kept in a map in `prepStmtNamespace` so that
`prepStmtNamespace` can implement `commit` and `rewind` methods.

Release note (performance improvement): Prepared statements are now more
efficiently cached.

#### prep: decrement statement reference counts in Cache.Init

Prepared statement reference counts are now decremented in `Cache.Init`
if the cache is non-empty. This makes it easier and safe to clear a
`Cache` with `Init`.

Also, some minor changes have been made to simplify the test oracle.

Release note: None

mgartner committed:
querycache,prep: move PrepareMetadata to new prep package

This is a mechanical change that moves `querycache.PrepareMetadata` to a
new `prep` package in preparation for future changes.

Release note: None

mgartner committed:
sql: replace prepared statement maps with prep.Cache

Four maps using to implement the prepared statements cache have been
repalced with `prep.Cache`. Additional, a snapshot of a connection's
portals is now kept in a map in `prepStmtNamespace` so that
`prepStmtNamespace` can implement `commit` and `rewind` methods.

Release note (performance improvement): Prepared statements are now more
efficiently cached.

mgartner committed:
sql,prep: move PreparedStatement to prep package

This is a mechanical change the moves `sql.PreparedStatement` to the
`prep` package in preparation for future changes.

Release note: None

mgartner committed:
prep: add Cache

A new `prep.Cache` data structure has been added which maps prepared
statement names to `*prep.Statement`s. `Cache` provides a concise API
and it is optimized for the most common interactions in most workloads:

  1. Fetching prepared statements with `Get`.
  2. Snapshotting the `Cache` with `Commit` when no changes have been
     applied since the last `Commit`.

In order to facilitate automatic transaction retries which rewind the
command buffer, changes to the prepared statements cache must be undone.
We currently achieve this using two maps: one up-to-date map and the
other a snapshot of the prepared statements after the last transaction
committed. Before retrying, we clear the up-to-date copy of the map and
copy the contents of the snapshot into it, effectively undoing and
changes to the cache in the current transaction. If the transaction is
successful, we clear the snapshot and copy the up-to-date map into it,
effectively creating a new snapshot.

This approach is expensive even if the prepared statements cache is not
altered. If the cache hasn't changed, we can avoid clearing and copying
into the snapshot, but we still need to check for map equality to
determine if there have been changes. This equality check has time
complexity proportional to the number of cached prepared statements.

Also, there are two other maps we use implement an LRU eviction policy.
They keep track of the ordering of most recent access for each statement
(one is up-to-date and the other is a snapshot). Every time a prepared
statement is executed that was not the previously executed statement,
the most recent access ordering changes. Therefore, we can rarely avoid
the clear and copy of the snapshot like we can with the other maps.

The current implementation also lacks a clear API - the four maps are
separate objects and directly accessed across the `sql` pacakge and
prone to misuse.

`Cache` achieves similar behavior more efficiently with a clear API. It
provides `Add`, `Get`, `Remove`, and other methods with obvious purpose.
The `Rewind` and `Commit` methods facilitate automatic retries. `Commit`
is a constant-time operation if there have been no changes to the
`Cache` since the last `Commit`. See the inline comments for more
details.

The behavior of `Cache`, when used, will vary slightly from the current
behavior. Prior to this commit prepared statements were evicted as soon
new statements were added. Now, prepared statements are evicted when the
cache is committed, i.e., when statements succeed (or fail without a
retriable error) and we know that the command buffer will not be
rewound. This difference in behavior should have no meaningful impact
for users.

Fixes #143732

Release note (performance improvement): Prepared statements are now more
efficiently cached.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144021)
<!-- Reviewable:end -->

mgartner commented:
Some micro-benchmark results:

```
name                                  old time/op    new time/op    delta
Sysbench/SQL/3node/oltp_point_select     176µs ± 6%     156µs ± 5%  -11.75%  (p=0.008 n=5+5)

name                                  old errs/op    new errs/op    delta
Sysbench/SQL/3node/oltp_point_select      0.00           0.00          ~     (all equal)

name                                  old alloc/op   new alloc/op   delta
Sysbench/SQL/3node/oltp_point_select    25.6kB ± 1%    25.2kB ± 1%   -1.45%  (p=0.008 n=5+5)

name                                  old allocs/op  new allocs/op  delta
Sysbench/SQL/3node/oltp_point_select       218 ± 1%       217 ± 0%     ~     (p=0.095 n=5+4)
```

I'll collect more benchmark results soon.

mgartner committed:
prep: decrement statement reference counts in Cache.Init

Prepared statement reference counts are now decremented in `Cache.Init`
if the cache is non-empty. This makes it easier and safe to clear a
`Cache` with `Init`.

Also, some minor changes have been made to simplify the test oracle.

Release note: None

mgartner commented:
Unfortunately, I could not measure any improvement in throughput in a full oltp_read_write test.

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice refactor!

> Reviewed 11 of 12 files at r1, 1 of 1 files at r2, 16 of 16 files at r3, 3 of 3 files at r4, 10 of 10 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144021#-:-ONwg4GdEuups5VRERV7:bnaorjf)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @angles-n-daemons and @rytaft)

___
*[`pkg/sql/logictest/testdata/logic_test/prepare` line 1653 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/144021#-ONweejC3Lhdpvka6hz5:-ONweejD3Lh9IiBJ2xV2:b-ori6kt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced80eea46d152622b93842abcfe823d3ed85e2c/pkg/sql/logictest/testdata/logic_test/prepare#L1653)):*
> ```
> 2  pscs11
> 2  pscs14
> 2  pscs15
> ```

Was it incorrect before? Why weren't these rows there before the change?
___
*[`pkg/sql/conn_executor.go` line 2019 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/144021#-ONwfd4X0jrRQOGmMXG7:-ONwfd4X0jrRQOGmMXG8:b-l8gzm4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4afeb3343d51302b99cf241d9bc4a8dbc4f982/pkg/sql/conn_executor.go#L2019)):*
> ```Go
> }
> 
> // reset empties the prepStmtNamespace.
> ```

nit: this comment should say `clear`.
___
*[`pkg/sql/conn_executor.go` line 2024 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/144021#-ONwfS5a1FuOdnVVC89v:-ONwfS5a1FuOdnVVC89w:bnogb9s) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4afeb3343d51302b99cf241d9bc4a8dbc4f982/pkg/sql/conn_executor.go#L2024)):*
> ```Go
> ) {
> 	// Commit the prepared statements cache to ensure that any statements that
> 	// have been removed but not committed are not leaked.
> ```

nit: comment is stale.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144021)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @angles-n-daemons, @DrewKimball, and @rytaft)

___
*[`pkg/sql/conn_executor.go` line 2019 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/144021#-ONwfd4X0jrRQOGmMXG7:-OO9fcB_Dxw_pGntHVTB:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4afeb3343d51302b99cf241d9bc4a8dbc4f982/pkg/sql/conn_executor.go#L2019)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

nit: this comment should say `clear`.
</blockquote></details>

done
___
*[`pkg/sql/conn_executor.go` line 2024 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/144021#-ONwfS5a1FuOdnVVC89v:-OO9gt447c3d4KQ6wiSR:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d4afeb3343d51302b99cf241d9bc4a8dbc4f982/pkg/sql/conn_executor.go#L2024)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

nit: comment is stale.
</blockquote></details>

done
___
*[`pkg/sql/logictest/testdata/logic_test/prepare` line 1653 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/144021#-ONweejC3Lhdpvka6hz5:-OO9h0mtA-_6lPCdxFE-:b-13cg3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/ced80eea46d152622b93842abcfe823d3ed85e2c/pkg/sql/logictest/testdata/logic_test/prepare#L1653)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Was it incorrect before? Why weren't these rows there before the change?
</blockquote></details>

There's a slight behavioral difference with this PR. Prior to this commit prepared statements were evicted as soon new prepared statements were added, e.g., in the middle of the transaction above. Now, prepared statements are evicted when the cache is committed, i.e., the statements have succeeded (or failed without a retriable error) and the command buffer no longer needs to be rewound. 

These new rows show the change in behavior. Rows with `which=2` show that nothing is evicted within the transaction. Rows with `which=3` show that there have been some evictions one the transaction commits.

I've updated the 3rd commit message to clarify this change in behavior. I've also added another explanatory sentence to this test's comment.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
> Reviewed 22 of 22 files at r7, 1 of 10 files at r10, 1 of 3 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144021)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @angles-n-daemons and @rytaft)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTR!

bors r+
